Simplified middle angle in Quaternion.to_euler for asymmetric sequences

evbernardes opened the pull request:
#### Brief description of what is fixed or changed
Two simplifications are made for the option `avoid_square_root=True`.
First, I put a call to `simplify.factor` inside the formula for the middle angle, so that running:
```python
Quaternion(a, b, c, d, norm=1).to_euler('zyx', avoid_square_root=True)[1]
```
Gives this:
```math
\operatorname{acos}{\left( 2 \, (b d - a c) \right)}- \frac{\pi}{2}
```
Instead of this:
```math
\operatorname{acos}{\left(\frac{\left(a - c\right)^{2}}{2} - \frac{\left(a + c\right)^{2}}{2} - \frac{\left(- b + d\right)^{2}}{2} + \frac{\left(b + d\right)^{2}}{2} \right)} - \frac{\pi}{2}
```

Second, I used the fact that $\cos(\theta + \pi / 2) = -\sin(\theta)$ to simplify $\operatorname{acos}(x) - \pi / 2 = \operatorname{asin}(- x)$ for asymmetric sequences, further simplifying the result to:
```math
\operatorname{asin}{\left(2 \left(a c - b d\right) \right)}
```

This has one big advantage: it is consistent with a formula for the pitch angle that is widely implemented.

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

evbernardes committed:
simplified middle angle in to_euler

evbernardes committed:
Added factor to symmetric for consistency

evbernardes committed:
minor edit

sylee957 commented on sympy/algebras/quaternion.py:
```diff
@@ -532,10 +534,16 @@ def to_euler(self, seq, angle_addition=True, avoid_square_root=False):
         if avoid_square_root:
-            n2 = self.norm()**2 if symmetric else 2 * self.norm()**2
-            angles[1] = acos((a*a + b*b - c*c - d*d) / n2)
+            if symmetric:
+                n2 = self.norm()**2
+                angles[1] = acos(factor((a * a + b * b - c * c - d * d) / n2))
+            else:
+                n2 = 2 * self.norm()**2
+                angles[1] = asin(factor((c * c + d * d - a * a - b * b) / n2))
```
I'm not sure if the `factor` has to be used than building the program with direct formula from symbolic `Quaternion(a, b, c, d)` value.

evbernardes replied:
What do you mean? That it's best to leave the `factor` out?

evbernardes replied:
Alternatively it can be a call to `simplify`, but I think it is usually best to avoid it and use something more direct instead, right?

sylee957 replied:
Anything about `Quaternion(expr1, expr2, expr3, expr4)` can be substituted from a single formula of 
`Quaternion(a, b, c, d).to_euler`

sylee957 replied:
So I think why not it's better to give result for arbitrary formula to be compatible with substitution. it can be an options to pick arbitrary `simplify, factor` to break it further down

evbernardes committed:
Removed internal calls to simplify.factor

sylee957 commented:
I think that you may need to add the test cases

evbernardes commented:
> I think that you may need to add the test cases

Since `avoid_square_root` is an option that was added later, when I implemented this one I also implemented these tests to test that both setting it to `False` (using regular formula) or `False` (using this option) would both give equivalent results.

Since the trigonometric simplifications aren't currently capable of simplifying square roots of arguments of inverse trigonometric functions, I tested this numerically, implementing this (testing every possible quadrant for every possible sequence):

```python
def test_to_euler_options():
    def test_one_case(q):
        angles1 = Matrix(q.to_euler(seq, True, True))
        angles2 = Matrix(q.to_euler(seq, False, False))
        angle_errors = simplify(angles1-angles2).evalf()
        for angle_error in angle_errors:
            # forcing angles to set {-pi, pi}
            angle_error = (angle_error + pi) % (2 * pi) - pi
            assert angle_error < 10e-7

    for xyz in ('xyz', 'XYZ'):
        for seq_tuple in permutations(xyz):
            for symmetric in (True, False):
                if symmetric:
                    seq = ''.join([seq_tuple[0], seq_tuple[1], seq_tuple[0]])
                else:
                    seq = ''.join(seq_tuple)

                for elements in product([-1, 0, 1], repeat=4):
                    q = Quaternion(*elements)
                    if not q.is_zero_quaternion():
                        test_one_case(q)
```

 In this case, I'm considering `avoid_square_root=False` as the ground truth, since it's already being tested by `test_to_euler`.

Adding this `asin` formula still passes both tests, so I think there is enough testing.
